chipflow_lib/cli.py

```diff
@@ -1,173 +1,75 @@
-                                }
-                            }
-                        },
-                    }
-                },
-            },
-        }
-    }
-}
-
+import inspect
+import sys
+import traceback
+import logging
 
-def _parse_config():
-    chipflow_root = _ensure_chipflow_root()
-    config_file = f"{chipflow_root}/chipflow.toml"
-    return _parse_config_file(config_file)
+from pprint import pformat
 
+from . import (
+    ChipFlowError,
+    _get_cls_by_reference,
+    _parse_config,
+)
+from .pin_lock import PinCommand
 
-def _parse_config_file(config_file):
-    with open(config_file, "rb") as f:
-        config_dict = tomli.load(f)
 
-    try:
-        jsonschema.validate(config_dict, config_schema)
-        return config_dict
-    except jsonschema.ValidationError as e:
-        raise ChipFlowError(f"Syntax error in `chipflow.toml` at `{'.'.join(e.path)}`: {e.message}")
+logging.basicConfig(stream=sys.stdout, level=logging.WARNING)
 
+class UnexpectedError(ChipFlowError):
+    pass
 
 def run(argv=sys.argv[1:]):
     config = _parse_config()
 
-    steps = {}
+    commands = {}
+    commands["pin"] = PinCommand(config)
+
     for step_name, step_reference in config["chipflow"]["steps"].items():
         step_cls = _get_cls_by_reference(step_reference, context=f"step `{step_name}`")
         try:
-            steps[step_name] = step_cls(config)
+            commands[step_name] = step_cls(config)
         except Exception:
             raise ChipFlowError(f"Encountered error while initializing step `{step_name}` "
                                 f"using `{step_reference}`")
 
     parser = argparse.ArgumentParser()
-    step_argument = parser.add_subparsers(dest="step", required=True)
-    for step_name, step_cls in steps.items():
-        step_subparser = step_argument.add_parser(step_name, help=inspect.getdoc(step_cls))
+
+    parser.add_argument(
+        "--verbose", "-v",
+        dest="log_level",
+        action="append_const",
+        const=10,
+    )
+
+    command_argument = parser.add_subparsers(dest="command", required=True)
+    for command_name, command in commands.items():
+        command_subparser = command_argument.add_parser(command_name, help=inspect.getdoc(command))
         try:
-            step_cls.build_cli_parser(step_subparser)
+            command.build_cli_parser(command_subparser)
         except Exception:
             raise ChipFlowError(f"Encountered error while building CLI argument parser for "
-                                f"step `{step_name}`")
+                                f"step `{command_name}`")
 
     args = parser.parse_args(argv)
-    try:
-        steps[args.step].run_cli(args)
-    except ChipFlowError:
-        raise
-    except Exception:
-        raise ChipFlowError(f"Encountered error while running CLI for step `{args.step}`")
-
+    if args.log_level:
+        log_level = max(logging.DEBUG, logging.WARNING - sum(args.log_level))
```

Comment:
hm, this seems a bit weird, what's the `sum` doing?

Reply:
Added commentary - its for multiple verbose flags